### PR TITLE
[KL-170] Add lock app period logic

### PR DESCRIPTION
**Pull request by thesvyatoslavtt**

### CODE REVIEW

File: apps/extension/background-script.ts [hunk 1]
```diff
@@ -38,26 +37,32 @@ scripting.registerContentScripts([
   }
 ]);
 
-runtime.onConnect.addListener(port => {
+const convertLockTime = (value: number) => value * 60 * 1000;
+
+async function fetchFromStorage<T = any>(key: string): Promise<T | null> {
```
Comment (Boltman92) [inline apps/extension/background-script.ts]: let's move convertLockTime and fetchFromStorage to utils

File: apps/extension/src/utils/background-script.utils.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { storage } from 'webextension-polyfill';
+
+export const convertLockTime = (value: number) => value * 60 * 1000;
+
+export async function fetchFromStorage<T = any>(key: string): Promise<T | null> {
```
Comment (unixvb) [inline apps/extension/src/utils/background-script.utils.ts]: you are using 2 different approaches:  
1. you are calling `set` method directly where you need that
2. you did create util for `get method`

follow single approach to work with `storage`

File: apps/extension/src/utils/background-script.utils.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  const items = await storage.local.get([key]);
```
Comment (unixvb) [inline apps/extension/src/utils/background-script.utils.ts]: you are using `storage.local` for storing password hash and session-related data, but you should use  `browser.storage.session`

File: apps/extension/background-script.ts [hunk 4]
```diff
@@ -20,10 +21,9 @@ import {
 import { createDAppNotificationResponse, getHexChanId } from './src/utils/network.utils';
 import { getState } from './src/utils/state.utils';
 
-const APP_LOCK_PERIOD = 5 * 60 * 1000;
-
-let passwordHash = INITIAL_PASSWORD_HASH;
-let lastUserActivityTimestamp = 0;
+const LOGIN_TOKEN = 'login_token';
+const LOCK_TIME_PERIOD = 'lock_time_period';
+const LAST_USER_ACTIVITY_TIMESTAMP = 'last_user_activity_timestamp';
```
Comment (unixvb) [inline apps/extension/background-script.ts]: add `_KEY` to the end of these variables.  (as this's the keys where data is stored)  
and rename `LOGIN_TOKEN` to `PASSWORD_HASH_KEY` - name should represent what data is stored there

File: libs/ui/src/screens/settings/screens/lock-time-selector/lock-time-selector.tsx [hunk 5]
```diff
@@ -1,17 +1,40 @@
...
 
 export const LockTimeSelector: FC = () => {
   const { goBack } = useNavigation();
+  const dispatch = useDispatch();
+  const lockTimePeriod = useLockTimePeriodSelector();
+
+  const onSelect = ({ value }: Option<number>) => {
+    if (isWeb) {
+      BackgroundMessager.setLockTimePeriod(value);
+      dispatch(setAppLockTimePeriod(value));
```
Comment (unixvb) [inline libs/ui/src/screens/settings/screens/lock-time-selector/lock-time-selector.tsx]: I don't like that we have 2 sources of truth here.  
Ideally is to use a single value from `browser.storage.session` in UI and Background scripts

File: apps/extension/background-script.ts [hunk 6]
```diff
@@ -251,14 +246,34 @@ runtime.onConnect.addListener(port => {
...
+        Date.now() > lastUserActivity + convertLockTime(lockTimePeriod ?? 1) &&
+        !isFullpageOpen &&
+        passwordHash !== null &&
+        passwordHash.length
+      ) {
+        storage.local.set({ [LOGIN_TOKEN]: INITIAL_PASSWORD_HASH, [LAST_USER_ACTIVITY_TIMESTAMP]: 0 });
+
+        return Promise.resolve('');
+      }
+
```
Comment (unixvb) [inline apps/extension/background-script.ts]: `background-script` should be as clean as possible.  

create `apps/extension/src/utils/session.utils.ts` and move all logic there.
Then `background-script` will look something like this 
```
runtime.onMessage.addListener(message => {
    switch (message.type) {
        case BackgroundMessageType.SetPasswordHash:
            return setSessionPaswordHash();
        case BackgroundMessageType.GetPasswordHash:
            return getSessionPasswordHash();
......
    }
})
```

File: apps/extension/background-script.ts [hunk 7]
```diff
@@ -38,26 +38,21 @@ scripting.registerContentScripts([
   }
 ]);
 
-runtime.onConnect.addListener(port => {
+runtime.onConnect.addListener(async port => {
```
Comment (unixvb) [inline apps/extension/background-script.ts:36]: why `async` now?

File: apps/extension/background-script.ts [hunk 8]
```diff
@@ -251,15 +246,16 @@ runtime.onConnect.addListener(port => {
 });
 
 // listen messages from UI
-runtime.onMessage.addListener((message: BackgroundMessage) => {
+runtime.onMessage.addListener(async (message: BackgroundMessage) => {
```
Comment (unixvb) [inline apps/extension/background-script.ts]: the same, why `async`?

File: apps/extension/src/utils/popup.utils.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import { windows, runtime, Windows } from 'webextension-polyfill';
```
Comment (unixvb) [inline apps/extension/src/utils/popup.utils.ts]: looks like `windows` utils, mb rename?

File: apps/extension/src/utils/session.utils.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import { INITIAL_PASSWORD_HASH } from '../../../../libs/ui/src/background-script';
+import { LAST_USER_ACTIVITY_TIMESTAMP_KEY, LOCK_TIME_PERIOD_KEY, PASSWORD_HASH_KEY } from '../constants/storage-keys';
+
+import { browserWithSession as browser } from './browser.utils';
+
+export const convertLockTime = (value: number) => value * 60 * 1000;
+
+export async function fetchFromStorage<T = any>(key: string): Promise<T | null> {
```
Comment (unixvb) [inline apps/extension/src/utils/session.utils.ts]: fix warning (removing any should work)

File: apps/extension/src/utils/session.utils.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+  const items = await browser.storage.session.get([key]);
+  if (key in items) {
+    return items[key];
+  }
+
+  return null;
+}
+
+export async function putToStorage<T = any>(item: Record<string, T>) {
```
Comment (unixvb) [inline apps/extension/src/utils/session.utils.ts]: fix warning

Comment (unixvb) [inline apps/extension/src/utils/session.utils.ts]: why this fn called `putToStorage` if you are calling `session.set` inside?

File: apps/extension/src/utils/session.utils.ts (hunk 10 above)
Comment (unixvb) [inline apps/extension/src/utils/session.utils.ts]: the same here, why `fetch` if you are calling `session.get`?

File: libs/ui/src/store/settings/settings.state.ts [hunk 12]
```diff
@@ -6,10 +8,12 @@ export interface SettingsState {
   isAnalyticsEnabled: boolean;
   isBiometricEnabled: boolean;
   showLoader: boolean;
+  lockTimePeriod: number;
```
Comment (unixvb) [inline libs/ui/src/store/settings/settings.state.ts:11]: why don't you use the value from `session storage`, but save data in 2 places instead?

File: apps/extension/background-script.ts [hunk 13]
```diff
@@ -38,26 +38,21 @@ scripting.registerContentScripts([
...
-
   // listen when UI is closed
   if (port.name === 'klaytn_wallet_ui') {
     port.onDisconnect.addListener(port => {
       if (isFullpagePort(port)) {
         isFullpageOpen = false;
       }
 
-      return (lastUserActivityTimestamp = Date.now());
+      return putToStorage({ [LAST_USER_ACTIVITY_TIMESTAMP_KEY]: Date.now() });
```
Comment (unixvb) [inline apps/extension/background-script.ts]: take a look at this line, `setSessionPaswordHash` and `setLockTimePeriod`.  
  
1. you are using different approaches to the same functionality.
2. these functions have identical code - that means, something wrong with it

**Review by unixvb — APPROVED** [review]

**Review by Boltman92 — APPROVED** [review]

**Review by yarmoshevych — APPROVED** [review]

**Pull request merged**